### PR TITLE
Feature/data retrieval

### PR DESCRIPTION
**Pull request by SolomonShorser-OICR**
This PR introduces 2 new main pieces of functionality to release-common-lib:
 - File Retrieval
 - A Callable object that unzips files

The File Retrieval code (and any classes it depends on) was mostly just copied from AddLinks. There is also some code that is specific to retrieving files from COSMIC because it is common to two separate projects: AddLinks and COSMICUpdate.

### CODE REVIEW

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+
+public class FileRetriever implements DataRetriever {
+
+	protected URI uri;
+	protected String destination;
+	protected Duration maxAge;
+	protected Duration timeout = Duration.ofSeconds(30);
+	protected int numRetries = 1;
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:40]: Any reason why it will only retry once? For other downloaders (I think AddLinks) I believe they retry more times than that

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+		//Before fetching anything, we need to check to see if the file already exists.
+		Path pathToFile = Paths.get(this.destination);
+		if (Files.exists(pathToFile))
+		{
+			BasicFileAttributes attributes = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+			
+			Instant fileCreateTime = attributes.creationTime().toInstant();
+			Instant now = Instant.now();
+			//If the file is older than the maxAge...
+			if (fileCreateTime.isBefore( now.minus(this.maxAge) ))
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:83]: Will maxAge be configurable or is there a hard-coded value somewhere?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+			downloadData();
+		}
+		
+		// Print some basic file stats.
+		if (Files.exists(pathToFile))
+		{
+			if (Files.isReadable(pathToFile))
+			{
+				BasicFileAttributes attribs = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+				logger.info("File Info: Name: {}, Size: {}, Created: {}, Modified: {}",this.destination,attribs.size(), attribs.creationTime(), attribs.lastModifiedTime());
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:108]: This is excellent because at some point in the future building a 'file download validator' for steps like addlinks and orthopairs could maybe use this

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+		else
+		{
+			// If something failed during the data retrieval, the file might not exist. If that happens, let the user know.
+			logger.error("File \"{}\" still does not exist after executing the file retriever!", pathToFile);
+		}
+	}
+
+	protected void downloadData() throws Exception
+	{
+		logger.trace("Scheme is: {}", this.uri.getScheme());
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:124]: I'm not familiar with this method, so might be a silly question -- any chance that a scheme wouldn't be in a uri?

Comment (cookersjs) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:124]: Oh nevermind, its just the protocol prefix of the url

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+				doFtpDownload();
+			}
+			else
+			{
+				throw new UnsupportedSchemeException("URI "+this.uri.toString()+" uses an unsupported scheme: "+this.uri.getScheme());
+			}
+		}
+		catch (URISyntaxException e)
+		{
+			logger.error("Error creating for download destination: " + this.destination, e);
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java]: typo - the 'for' can be removed

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+			}
+			catch (ConnectTimeoutException e)
+			{
+				// we will only be retrying the connection timeouts, defined as the time required to establish a connection.
+				// we will not handle socket timeouts (inactivity that occurs after the connection has been established).
+				// we will not handle connection manager timeouts (time waiting for connection manager or connection pool).
+				e.printStackTrace();
+				logger.info("Failed due to ConnectTimeout, but will retry {} more time(s).", retries);
+				retries--;
+				done = retries + 1 <= 0;
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:292]: Any reason why this is evaluated again, and within the catch block?

File: src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		// 1) Generate a base64-encoded string of the username and password.
+		// 2) Send this encoded string in the "Authorization: Basic" header to the URL of the file you want.
+		// 3) Parse the response - extract the "url" JSON attribute and then download THAT URL to get the REAL file.
+		
+		// Encoded string.
+		String encodedUsernamePassword = Base64.getEncoder().encodeToString((this.userName + ":" + this.password).getBytes("UTF-8"));
+		
+		HttpGet get = new HttpGet(this.uri);
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		int delayInMilliseconds = 1000 * (int)this.timeout.getSeconds();
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java:61]: I'm a bit confused - the `getSeconds()` either returns seconds, and so would need to be _divided_ by 1000 to get it in milliseconds, or it returns milliseconds and so would need to be _multiplied_ by 1000 to get seconds, and so the variable would need to be called `delayInSeconds`?

Comment (cookersjs) [reply src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java:61]: Oh wow nevermind I think I need a coffee

File: src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+		catch (IOException e)
+		{
+			logger.error("IOException was caught, probably caused by some network communication issue. Message: {}", e.getMessage());
+			e.printStackTrace();
+		}
+		catch (URISyntaxException e)
+		{
+			logger.error("The URL from COSMIC might be malformed. URL is: \"{}\", Error message is: {}", downloadURL, e.getMessage() );
+			e.printStackTrace();
+		}
```
Comment (cookersjs) [inline src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java:104]: I think one of the recent releases had COSMIC require the owner of the account to 'verify' their account on the actual cosmic website. I couldn't download the file because authentication wasn't working programmatically, and so after getting the credentials I logged in on the website... at which point the programmatic file downloads worked. Any way to guard against that type of situation? I can't recreate it at the moment since it was just something that they seemed to do after a set amount of time (I'm guessing like a year)

File: src/main/java/org/reactome/util/general/GUnzipCallable.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.file.Path;
+import java.util.concurrent.Callable;
+import java.util.zip.GZIPInputStream;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * An object that can unzip a GZIP file, and mutliple instances of this type could be run in parallel.
```
Comment (cookersjs) [inline src/main/java/org/reactome/util/general/GUnzipCallable.java]: typo, mutliple

File: src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+		// to run this test, you will need to create your own auth.properties under src/test/resources
+		// and populate it with a valid username and password for COSMIC. 
+		Properties props = new Properties();
+		props.load(new FileReader("src/test/resources/auth.properties"));
+		COSMICFileRetriever retriever = new COSMICFileRetriever();
+		retriever.setUserName(props.getProperty("username"));
+		retriever.setPassword(props.getProperty("password"));
+		
+		// classification.csv is a pretty small file, ~2MB I think, so a good choice for testing downloads.
+		retriever.setDataURL(new URI("https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v92/classification.csv"));
```
Comment (cookersjs) [inline src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java:34]: Just commenting to point out the v92 in the string - I know there isn't much that can be done about it as far as I know, but would be good to at least try in the future to have it worked with the current release, or the release that we download the cosmic files from. Could maybe be put as a properties? (not this PR, just thinking)

**Review by cookersjs — DISMISSED** [review]
No blockers, just some questions and comments. This is a great feature for eventually unifying our file download steps

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 1 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:40]: This is a default value. There is a `setNumRetries` method to change this, if neded.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 2 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:83]: Yes, there is a `setMaxAge` method.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 6 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:292]: I'm not sure I understand the issue here.

File: src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java (hunk 8 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java:104]: I didn't even know that such a situation could occur! It seems like an issue very specific to COSMIC. If I know what HTTP response code and headers they sent back in this situation, and could easily test any code I wrote for it, it _might_ be possible, but it seems like a pretty rare occurrence to need to program for.

File: src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java (hunk 10 above)
Comment (SolomonShorser-OICR) [reply src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java:34]: Maybe, though this is testing code so it really just cares that some download works.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 6 above)
Comment (cookersjs) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:292]: Ah nevermind, I think I was confused about the boolean being updated/checked within the catch block. Looking over it again this morning, I get it

File: src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java (hunk 8 above)
Comment (cookersjs) [reply src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java:104]: Yea, I wouldn't worry about doing anything else - its an easy fix and I don't know how we'd recreate it anyways

File: src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java (hunk 10 above)
Comment (cookersjs) [reply src/test/java/org/reactome/release/common/dataretrieval/COSMICFileRetrieverIT.java:34]: Fair enough, wasn't sure if we would at some point want it to focus on the current release, but it definitely doesn't matter at this point in time

File: src/main/java/org/reactome/release/common/CustomLoggable.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.logging.log4j.core.appender.FileAppender;
+import org.apache.logging.log4j.core.appender.RollingRandomAccessFileAppender;
+import org.apache.logging.log4j.core.appender.rolling.RollingRandomAccessFileManager;
+import org.apache.logging.log4j.core.appender.rolling.RolloverStrategy;
+import org.apache.logging.log4j.core.appender.rolling.TriggeringPolicy;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.LoggerConfig;
+
+/**
+ * An interface for classes that may want to customise they way they log.
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/CustomLoggable.java]: Typo - "**the** way they log"?

File: src/main/java/org/reactome/release/common/CustomLoggable.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+ *
+ */
+public interface CustomLoggable
+{
+	/**
+	 * Creates the logger. 
+	 * @param logFileName - custom name for log file.
+	 * @param oldAppenderName - the name of the appender to copy config from.
+	 * @param newAppenderName - the name of the new appender.
+	 * @param append - append, or not.
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/CustomLoggable.java]: The alternative is truncate?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+		this.setRetrieverName(retrieverName);
+		this.logger = this.createLogger(this.retrieverName, "RollingRandomAccessFile", this.getClass().getName(), true, Level.DEBUG, this.logger, "Data Retriever");
+	}
+	
+	@Override
+	public void fetchData() throws Exception 
+	{
+		if (this.uri == null)
+		{
+			throw new RuntimeException("You must provide a URI from which the file will be downloaded!");
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:68]: If you want, you could make this an IllegalStateException, so it is more specific

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+		}
+		else if (this.destination.trim().length() == 0)
+		{
+			throw new RuntimeException("You must provide a destination to which the file will be downloaded!");
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:72]: Same here: https://github.com/reactome/release-common-lib/pull/4/files#r579433624

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+
+	protected void downloadData() throws Exception
+	{
+		logger.trace("Scheme is: {}", this.uri.getScheme());
+		Path path = null;
+		try
+		{
+			path = Paths.get(this.destination);
+			Files.createDirectories(path.getParent());
+			if (this.uri.getScheme().equals("http") || this.uri.getScheme().equals("https"))
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:130]: If you want you could shorten this to be `this.uri.getScheme().startsWith("http")`

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 13 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:68]: For the next version of release-common-lib after this one, I was planning to put some time in for custom exception classes. The code is now an API that is getting more and more widely used so I think it's time to make that change. This would probably get fixed as a part of that.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+			{
+				
+				doHttpDownload(path);
+			}
+			else if (this.uri.getScheme().equals("ftp") || this.uri.getScheme().equals("sftp"))
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:135]: Similar here: `this.uri.getScheme().endsWith("ftp")` (though I'm okay leaving it as is in this case since I always forget that the "s" comes at the start for secure ftp!

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+	}
+
+	protected void doFtpDownload() throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		doFtpDownload(null, null);
+	}
+	
+	protected void doFtpDownload(String user, String password) throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		user = user == null || user.trim().equals("") ? "anonymous" : user;
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java]: Maybe:

```
if (user == null || user.trim().equals("")) {
    user = "anonymous"
}
```
I find I need to do a double-take to realize it's checking and resetting itself if necessary.

File: src/main/java/org/reactome/release/common/CustomLoggable.java (hunk 12 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/CustomLoggable.java]: I believe so.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+		password = password == null || password.trim().equals("") ? "" : password;
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java]: Same here: https://github.com/reactome/release-common-lib/pull/4/files#r579443467

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+				{
+					logger.error("No data returned from server for {}", this.uri.toString());
+				}
+			}
+		}
+		catch (IOException e)
+		{
+			logger.error("Error while retrieving the file: {}",e.getMessage());
+			e.printStackTrace();
+			throw new IOException(e);
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:210]: Is the print of the stacktrace needed if the exception is rethrown and handled later?  Also, the rethrow could be `throw e` since it's already an IOException that caught.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+		}
+		//Should probably have more/better reply-code checks.
+		logger.debug("retreive file reply code: {}",client.getReplyCode());
+		if (client.getReplyString().matches("^5\\d\\d.*") || (client.getReplyCode() >= 500 && client.getReplyCode() < 600) )
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:214]: Why the regex in addition to the check of between 500 and 600?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 17 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java]: Ew. What was I thinking?? :facepalm:

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+		{
+			String errorString = "5xx reply code detected (" + client.getReplyCode() + "), reply string is: "+client.getReplyString();
+			logger.error(errorString);
+			throw new Exception(errorString);
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:218]: Should/could this be a `RuntimeException`

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 19 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:210]: As mentioned earlier, exceptions are getting messy in this library. I'm planning to fix that up in a future release. Fixing everything that needs fixing is probably out of scope for version 1.2.1.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 20 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:214]: The regex is matched against replyString, the range is checked against replyCode. I imagine there was some situation where a server sent an error in one field but not the other, but this code is rather old so I can't recall the _exact_ situation.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		RequestConfig config = RequestConfig.copy(RequestConfig.DEFAULT)
+											.setConnectTimeout(1000 * (int)this.timeout.getSeconds())
+											.setSocketTimeout(1000 * (int)this.timeout.getSeconds())
+											.setConnectionRequestTimeout(1000 * (int)this.timeout.getSeconds()).build();
+		
+		get.setConfig(config);
+		
+		int retries = this.numRetries;
+		boolean done = retries + 1 <= 0;
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:261]: Why not `retries < 0`?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 21 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:218]: Please see my other comments on Exceptions.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 22 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:261]: It's been a while... I think it had something to do with the message printing out something like `...but will retry 0 more time(s)` right before it did the last retry (and stopping when it hit -1). It makes more sense if it says "will retry 1 more time" before that last retry. Are there other ways to do this? Yes, this is just one way.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+		while(!done)
```
Comment (jweiser) [inline src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:262]: Is it the case this block is always expected to run once (i.e. the number of retries should always be 0 or more)?  If so, it might be better to use a "do-while" loop so the condition is evaluated at the end and `boolean done` is declared outside the loop but only assigned inside.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 22 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:261]: Yes, but as a new reader I find it confusing that retries can potentially be -1.  I think it's better to refactor so retries can't be below 0:

```
boolean done;
do {
    ...
    // If retry not needed
    done = true;
    // If retry needed
    if (retries == 0) {
        done = true;
    } else {
        retries--;
    }
} while (!done);

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 23 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:262]: I like while-loops because you can read the condition at the top. I find it preferable to looking for the end of the loop to understand the loop condition.

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+			String dest = "/tmp/testFetchData_"+ String.valueOf((new Random()).nextInt(Integer.MAX_VALUE));
+			retriever.setFetchDestination(dest);
+			Duration age = Duration.of(5,ChronoUnit.SECONDS);
+			retriever.setMaxAge(age);
+			
+			retriever.fetchData();
+			assertTrue(Files.exists(Paths.get(dest)));
+			
+			//Sleep for two seconds, and then re-download because the file is stale
+			Thread.sleep(Duration.of(2,ChronoUnit.SECONDS).toMillis());
```
Comment (jweiser) [inline src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java]: Will this guarantee the file is stale since the max age is set to 5 seconds and then the sleep is for 2 seconds?

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+			retriever.fetchData();
+			assertTrue(Files.exists(Paths.get(dest)));
+			//now set a longer maxAge.
+			age = Duration.of(100,ChronoUnit.SECONDS);
+			retriever.setMaxAge(age);
+			// this time, the file will not be stale (because maxAge is larger) so nothing will be downloaded.
+			retriever.fetchData();
+			//check that the file exists.
+			assertTrue(Files.exists(Paths.get(dest)));
```
Comment (jweiser) [inline src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java:110]: Should the tests delete the files after they are downloaded and the test is complete so there is a clean state if they are re-run?

File: src/test/java/org/reactome/util/general/GUnzipCallableTest.java [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+		catch (InterruptedException e)
+		{
+			e.printStackTrace();
+			fail();
+		}
+		catch (IOException e)
+		{
+			e.printStackTrace();
+			fail();
+		}
```
Comment (jweiser) [inline src/test/java/org/reactome/util/general/GUnzipCallableTest.java]: You could combine this as:
```
} catch (InterruptedException | IOException e) {
     e.printStackTrace();
     fail();
}
```

File: src/main/java/org/reactome/release/common/CustomLoggable.java (hunk 12 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/CustomLoggable.java]: It might be good to state that explicitly in JavaDoc comment

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 13 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:68]: Okay, as long as it's on the radar.  Would it be before or after this PR is merged?  My concern is the comments getting lost showing the places for reworking if the changes are after this PR is merged

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 23 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:262]: I think it might be warranted to a do loop in this case (or a while loop with an internal break) because the loop condition is changed based on a successful download or running out of retries which requires looking in the loop body to understand since the condition is stored in the `done` variable which is set outside and altered inside rather than described in the while statement at the top itself.

In any case, I find it confusing that retries becomes -1 to end the loop.  Intuitively, I would expect that retries would be 0 or more.  How the loop is structured is up to you, but that condition really makes confused with the code and would be better if it could exit the loop when it finds the download failed and the retries are 0.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 13 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:68]: After this PR is merged. _This_ PR is too big, already.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 23 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:262]: I may consider making some changes...

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java (hunk 25 above)
Comment (SolomonShorser-OICR) [reply src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java:110]: I think I specifically didn't delete them so I could manually verify them, outside of the test.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 13 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:68]: Okay, you could just refer back to this PR for the comments to see what was suggested if needed.

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java (hunk 25 above)
Comment (jweiser) [reply src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java:110]: Okay, but this might interfere with re-running the tests?  Maybe do a delete if exists before the tests begin then?

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java (hunk 24 above)
Comment (SolomonShorser-OICR) [reply src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java]: Old commit wasn't checked in...

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java (hunk 25 above)
Comment (SolomonShorser-OICR) [reply src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java:110]: Interference seems unlikely, since the filenames include random numbers.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 23 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:262]: I think rewriting this code is out of scope for this PR - it is old, existing code (from AddLinks) that works fine. Changes may be made in the future but I think the effort to change and test would go beyond the scope for this PR which was to integrate the code.

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java [hunk 27]
```diff
@@ -91,18 +91,18 @@ public void testFetchData() throws Exception
...
-			Duration age = Duration.of(5,ChronoUnit.SECONDS);
+			Duration age = Duration.of(5, ChronoUnit.SECONDS);
 			retriever.setMaxAge(age);
 			
 			retriever.fetchData();
 			assertTrue(Files.exists(Paths.get(dest)));
 			
-			//Sleep for two seconds, and then re-download because the file is stale
-			Thread.sleep(Duration.of(2,ChronoUnit.SECONDS).toMillis());
+			//Sleep for two seconds, and then re-download because the file is stale.
```
Comment (jweiser) [inline src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java]: "Two" should be changed to "Six" to be in sync with the code?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 5 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java]: Does this still need to be addressed?

File: src/main/java/org/reactome/util/general/GUnzipCallable.java (hunk 9 above)
Comment (jweiser) [reply src/main/java/org/reactome/util/general/GUnzipCallable.java]: Does this still need to be addressed?

File: src/main/java/org/reactome/release/common/CustomLoggable.java (hunk 11 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/CustomLoggable.java]: Does this still need to be addressed?

File: src/main/java/org/reactome/release/common/CustomLoggable.java (hunk 12 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/CustomLoggable.java]: Does this still need to be addressed?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 15 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:130]: Any thoughts on this?

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 16 above)
Comment (jweiser) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:135]: Any thoughts on this?

File: src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java (hunk 27 above)
Comment (SolomonShorser-OICR) [reply src/test/java/org/reactome/release/common/dataretrieval/TestFileRetrieval.java]: :facepalm:

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 16 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:135]: It works as it is. May change later but not interested in re-testing right now.

File: src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java (hunk 15 above)
Comment (SolomonShorser-OICR) [reply src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java:130]: It works as it is. May change later but not interested in re-testing right now.

**Review by jweiser — APPROVED** [review]
